sw/device/tock/chips/opentitan_common/src/uart.rs

```diff
@@ -0,0 +1,320 @@
+        regs.intr_enable.modify(intr::rx_watermark::SET);
+        regs.fifo_ctrl.write(fifo_ctrl::rxilvl.val(0 as u32));
+    }
+
+    fn disable_rx_interrupt(&self) {
+        let regs = self.registers;
+
+        // Generate an interrupt if we get any value in the RX buffer
+        regs.intr_enable.modify(intr::rx_watermark::CLEAR);
+
+        // Clear the interrupt bit (by writing 1), if it happens to be set
+        regs.intr_state.write(intr::rx_watermark::SET);
+    }
+
+    fn tx_progress(&self) {
+        let regs = self.registers;
+        let idx = self.tx_index.get();
+        let len = self.tx_len.get();
+
+        if idx < len {
+            self.tx_buffer.map(|tx_buf| {
+                let tx_len = len - idx;
+
+                for i in 0..tx_len {
+                    if regs.status.is_set(status::txfull) {
+                        break;
+                    }
+                    let tx_idx = idx + i;
+                    regs.wdata.write(wdata::data.val(tx_buf[tx_idx] as u32));
+                    self.tx_index.set(tx_idx + 1)
+                }
+            });
+
+            // If the FIFO started empty above, when the first character was written, it would have
+            // been immediately dispatched to the shift register, latching the TX_EMPTY state.
+            // That should be cleared before enabling the interrupt, so a subsequent empty state
+            // will latch the state transition again.
+            regs.intr_state.write(intr::tx_empty::SET);
+
+            // With data queued, (re)enable the TX interrupt if the FIFO is not empty
+            if !regs.status.is_set(status::txempty) {
+                self.enable_tx_interrupt();
+                if !regs.status.is_set(status::txempty) {
+                    // Assuming the FIFO stayed non-empty while the interrupt was being enabled,
+                    // our work here is done
+                    return;
+                } else {
+                    // Otherwise, turn off the interrupt and handle below
+                    self.disable_tx_interrupt();
+                }
+            }
+        }
+
+        if self.tx_index.get() == self.tx_len.get() {
+            self.tx_client.map(|client| {
+                self.tx_buffer.take().map(|tx_buf| {
+                    client.transmitted_buffer(tx_buf, self.tx_len.get(), ReturnCode::SUCCESS);
+                });
+            });
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let intrs = regs.intr_state.extract();
+
+        if intrs.is_set(intr::tx_empty) {
+            self.disable_tx_interrupt();
+            self.tx_progress();
+        } else if intrs.is_set(intr::rx_watermark) {
+            self.disable_rx_interrupt();
+            // TODO: real RX processing
+        }
+    }
+
+    pub fn transmit_sync(&self, bytes: &[u8]) {
+        let regs = self.registers;
+        for b in bytes.iter() {
```

Comment:
You can get an iterator of type `u8` instead of `&u8` by adding `.copied()`, which copies each iterated element.